File: src/lib/core/http.ts
// Small shared HTTP helpers for provider adapters: timeout + retry/backoff on
// 429 and 5xx (honoring Retry-After). Kept dependency-free so every adapter —
// in the Next.js app and the standalone worker — can use it.

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export interface RetryOptions {
  attempts?: number;
  timeoutMs?: number;
  backoffMs?: number[];
}

/**
 * POST/GET with retry + backoff on transient failures. Throws on final failure
 * (non-2xx that isn't retryable, or exhausted retries).
 */
export async function fetchWithRetry(
  url: string,
  init: RequestInit = {},
  opts: RetryOptions = {}
): Promise<Response> {
  const attempts = opts.attempts ?? 4;
  const timeoutMs = opts.timeoutMs ?? 20000;
  const backoff = opts.backoffMs ?? [1000, 4000, 10000, 20000];
  let lastErr = "";

  for (let i = 0; i < attempts; i++) {
    try {
      const res = await fetch(url, { ...init, signal: AbortSignal.timeout(timeoutMs) });
      if (res.ok) return res;
      lastErr = `${res.status} ${res.statusText}`;
      if (res.status === 429 || res.status >= 500) {
        const retryAfter = Number(res.headers.get("retry-after")) * 1000;
        if (i < attempts - 1) await sleep(retryAfter > 0 ? retryAfter : backoff[i]);
        continue;
      }
      // Non-retryable: surface server detail to ease debugging.
      const detail = await res.text().catch(() => "");
      throw new Error(`${lastErr}${detail ? ` — ${detail.slice(0, 300)}` : ""}`);
    } catch (e) {
      lastErr = (e as Error).message;
      if (i < attempts - 1) await sleep(backoff[i]);
    }
  }
  throw new Error(lastErr || "request failed");
}

File: worker_py/http.py
"""Async HTTP with retry/backoff on 429 + 5xx (honoring Retry-After).

Port of ``src/lib/core/http.ts`` (``fetchWithRetry``). Kept dependency-light so
every provider adapter can use it.
"""

import asyncio
from typing import Optional

import httpx

DEFAULT_BACKOFF = [1.0, 4.0, 10.0, 20.0]

_client: Optional[httpx.AsyncClient] = None


def _get_client() -> httpx.AsyncClient:
    global _client
    if _client is None:
        _client = httpx.AsyncClient()
    return _client


async def request_with_retry(
    method: str,
    url: str,
    *,
    headers: Optional[dict] = None,
    json: Optional[dict] = None,
    attempts: int = 4,
    timeout_ms: int = 20000,
    backoff=None,
) -> httpx.Response:
    """Issue a request, retrying transient failures. Raises on final failure."""
    backoff = backoff or DEFAULT_BACKOFF
    timeout = httpx.Timeout(timeout_ms / 1000)
    client = _get_client()
    last_err = ""

    for i in range(attempts):
        try:
            res = await client.request(method, url, headers=headers, json=json, timeout=timeout)
            if res.is_success:
                return res
            last_err = f"{res.status_code} {res.reason_phrase}"
            if res.status_code == 429 or res.status_code >= 500:
                retry_after = 0.0
                try:
                    retry_after = float(res.headers.get("retry-after", "0"))
                except ValueError:
                    retry_after = 0.0
                if i < attempts - 1:
                    await asyncio.sleep(retry_after if retry_after > 0 else backoff[min(i, len(backoff) - 1)])
                continue
            detail = ""
            try:
                detail = res.text
            except Exception:
                detail = ""
            raise RuntimeError(f"{last_err}{(' — ' + detail[:300]) if detail else ''}")
        except Exception as e:  # noqa: BLE001 — mirror the JS catch-all retry
            last_err = str(e)
            if i < attempts - 1:
                await asyncio.sleep(backoff[min(i, len(backoff) - 1)])

    raise RuntimeError(last_err or "request failed")
